docs/content/concepts/configuration/advanced-config-types.mdx

```diff
@@ -0,0 +1,370 @@
+from typing import List, Dict
+
+class MyDataStructuresConfig(Config):
+    user_names: List[str]
+    user_scores: Dict[str, int]
+
+@asset
+def scoreboard(config: MyDataStructuresConfig):
+    ...
+
+result = materialize(
+    [scoreboard],
+    run_config=RunConfig(
+        {
+            "scoreboard": MyDataStructuresConfig(
+                user_names=["Alice", "Bob"],
+                user_scores={"Alice": 10, "Bob": 20},
+            )
+        }
+    ),
+)
+```
+
+---
+
+## Nested schemas
+
+Schemas can be nested in one another, or in basic Python data structures.
+
+Here, we define a schema which contains a mapping of user names to complex user data objects.
+
+```python file=/guides/dagster/pythonic_config/pythonic_config.py startafter=start_nested_schema_config endbefore=end_nested_schema_config dedent=4
+from dagster import asset, materialize, Config, RunConfig
+from typing import Dict
+
+class UserData(Config):
+    age: int
+    email: str
+    profile_picture_url: str
+
+class MyNestedConfig(Config):
+    user_data: Dict[str, UserData]
+
+@asset
+def average_age(config: MyNestedConfig):
+    ...
+
+result = materialize(
+    [average_age],
+    run_config=RunConfig(
+        {
+            "average_age": MyNestedConfig(
+                user_data={
+                    "Alice": UserData(age=10, email="alice@gmail.com", profile_picture_url=...),
+                    "Bob": UserData(age=20, email="bob@gmail.com", profile_picture_url=...),
+                }
+            )
+        }
+    ),
+)
+```
+
+---
+
+## Permissive schemas
+
+By default, `Config` schemas are strict, meaning that they will only accept fields that are explicitly defined in the schema. This can be cumbersome if you want to allow users to specify arbitrary fields in their config. For this purpose, you can use the `PermissiveConfig` base class, which allows arbitrary fields to be specified in the config.
+
+```python file=/guides/dagster/pythonic_config/pythonic_config.py startafter=start_permissive_schema_config endbefore=end_permissive_schema_config dedent=4
+from dagster import asset, PermissiveConfig
+from typing import Optional
+import requests
+
+class FilterConfig(PermissiveConfig):
+    title: Optional[str] = None
+    description: Optional[str] = None
+
+@asset
+def filtered_listings(config: FilterConfig):
+    # extract all config fields, including those not defined in the schema
+    url_params = config.dict()
+    return requests.get("https://my-api.com/listings", params=url_params).json()
+
+filtered_listings(FilterConfig(title="hotel", beds=4))
+```
+
+---
+
+## Union types
+
+Union types are supported using Pydantic [discriminated unions](https://docs.pydantic.dev/usage/types/#discriminated-unions-aka-tagged-unions). Each union type must be a subclass of <PyObject object="Config"/>. The `discriminator` argument to <PyObject object="Field"/> specifies the field that will be used to determine which union type to use.
```

Comment:
I would not that in the legacy system this was the `Selector` class in order to orient people familiar with the old system